### PR TITLE
`BtcSwapTx`: update refund tx to consider all utxos

### LINKED ISSUE #72: Incoming chain swaps: refund only considers 1st UTXO
For incoming chain swaps (BTC -> LBTC), in case of address re-use (e.g. user locks up funds via several txs to the swap lockup address), the refund will only consider the 1st utxo. This means it's not possible to refund any of the subsequent utxos. It's also not possible to claim them, as the claim is also triggered for the 1st utxo.

There are several variants of this:
- user sends funds several times to the lockup address, then tries to refund
- user goes through a chain swap completely (which ends either with a claim or a refund), then sends funds again to the lockup address

In both cases, it would be useful if `BtcSwapTx::new_refund` would build a refund tx using all available utxos.

### PR DESCRIPTION
**Pull request by ok300**
This PR updates `BtcSwapTx` and its `new_refund` method to consider all available utxos.

Fixes #72

### CODE REVIEW

**Comment by ok300** [comment]
@i5hi as I'm still testing this, I marked it as a draft and wanted to ask for your high level thoughts and feedback:
- is this the proper way of addressing this, or do you see a better way?
- should the two `BtcSwapTx` fields (`utxo` and `utxos`) be merged into one?
- are there any other implications or considerations for this PR that I might have missed?

**Comment by i5hi** [comment]
The approach is good! 

We don't need the old `utxo` field anymore.

**Comment by ok300** [comment]
Alright, I continued with this approach and removed the `utxo` field in 3587e3c581f072282a6f3dbf88e2e3f4adfde3a1 .

Short summary of the changes:
- removed `utxo` field
- updated `new_refund` to use `fetch_utxos` instead of `fetch_utxo`
- updated `sign_claim` to rely only on the 1st utxo in `utxos` (*)
- updated `sign_refund`
  - cooperative branch: rely only on the 1st utxo in `utxos` (*)
  - non-cooperative branch: use all available utxos as inputs

I'm not fully sure about the parts marked with (*), more specifically:

* Can more than 1 utxos be claimed?
  * I assume no, because as soon as 1 utxo shows up on the lockup address, the Boltz Backend advances the swap to the next state.
* Can more than 1 utxos be refunded in cooperative mode?
  * I assume no, because this needs interaction with Boltz and intuitively it makes sense to only be able to refund what can be claimed (see above).

What do you think @i5hi? Are my assumptions correct?

**Comment by i5hi** [comment]
- Can more than 1 utxo be claimed?
Technically, yes but in practice, No. 

Boltz will never lockup more than 1 utxo, so wont need to. However, for consistency the code should attempt to spend all utxos, incase for whatever reason there is more than 1 - no harm I suppose.

- Can more than 1 utxo be coop refunded?
Yes. 

We refund what we lock* up - so incase we lock up more than 1 utxo, we must refund both.

**Comment by i5hi** [comment]
@michael1011 Need confirmation from you on this

**Comment by i5hi** [comment]
Infact, we should stick with your original approach of using`fetch_utxo` for `new_claim` and always expect and spend just 1 utxo.

while `new_refund` will use `fetch_utxos` and always attempt to spend all.

**Comment by michael1011** [comment]
> Can more than 1 utxos be claimed?

In theory, yes. But it's very likely never going to happen (unless there is a grave bug) that we lock more than one UTXO for a swap.

> Can more than 1 utxos be refunded in cooperative mode?

Yes. When asking for a cooperative refund signature, one of the parameters is `index` which is the number of the input you want to have a partial signature for. That way, you can do multiple cooperative refunds (for a single swap or multiple ones) in a single transaction. When we have a swap in a failed state, we are happy to create as many refund signatures as you would like.

File: src/swaps/bitcoin.rs [hunk 1]
```diff
@@ -400,27 +400,36 @@ impl BtcSwapScript {
...
-            let outpoint_0 = OutPoint::new(utxos[0].tx_hash, utxos[0].tx_pos as u32);
-            let txout = TxOut {
-                script_pubkey: spk,
-                value: Amount::from_sat(utxos[0].value),
-            };
-            Ok(Some((outpoint_0, txout)))
-        }
+    /// Fetch the first (utxo,amount) pair for the script_pubkey of this swap, which is typically
+    /// the lockup utxo. Returns None if no utxo for the script_pubkey is found.
+    pub fn fetch_utxo(
```
Comment (michael1011) [inline src/swaps/bitcoin.rs]: Should we get rid of this function completely?

File: src/swaps/bitcoin.rs [hunk 2]
```diff
@@ -979,7 +1011,7 @@ impl BtcSwapTx {
             )?;
 
             // Step 7: Get boltz's partial sig
-            let refund_tx_hex = serialize(&refund_tx).to_lower_hex_string();
+            let refund_tx_hex = serialize(&refund_tx_cooperative).to_lower_hex_string();
             let partial_sig_resp = match self.swap_script.swap_type {
                 SwapType::Chain => {
                     boltz_api.get_chain_partial_sig(&swap_id, &pub_nonce, &refund_tx_hex)
```
Comment (michael1011) [inline src/swaps/bitcoin.rs]: `get_chain_partial_sig` and `get_submarine_partial_sig` have `index` 0 hardcoded in them; we need to adjust those to have a parameter which needs to be the index of the input of the transaction

File: src/swaps/bitcoin.rs [hunk 3]
```diff
@@ -1036,48 +1068,55 @@ impl BtcSwapTx {
 
             let mut witness = Witness::new();
             witness.push(final_schnorr_sig.to_vec());
+            refund_tx_cooperative.input[0].witness = witness;
```
Comment (michael1011) [inline src/swaps/bitcoin.rs]: We need a witness for *all* inputs we want to spend

File: src/swaps/bitcoin.rs (hunk 1 above)
Comment (ok300) [reply src/swaps/bitcoin.rs]: Done: 82c6bfb526cba84cb6e1273dcb143b078b5be1ea

File: src/swaps/bitcoin.rs (hunk 2 above)
Comment (ok300) [reply src/swaps/bitcoin.rs]: Done: e3f9c573b132eceef96465bbd811a1cbea0cfc61

File: src/swaps/bitcoin.rs (hunk 3 above)
Comment (ok300) [reply src/swaps/bitcoin.rs]: Done: aff632f82e738b578eb618e7fce1ff91af78309d

One question I have here is this:

"Step 7: Get boltz's partial sig" starts by serializing the `refund_tx` to hex. It then calculates and attaches the witness for this input. This is inside a loop iterating over `refund_tx.input[input_index]`.

However this loop modifies the `refund_tx` at the end of each iteartion, by setting the witness for this input. So the tx hex will be different in each loop iteration. More specifically, it's this section: https://github.com/SatoshiPortal/boltz-rust/blob/e3f9c573b132eceef96465bbd811a1cbea0cfc61/src/swaps/bitcoin.rs#L995-L1060

I wonder, does the changing `refund_tx_hex` affect the partial sig calculations in Step 7, as they depend on this hex and are repeated for each input? Or does the partial sig logic account for this?

**Comment by ok300** [comment]
@i5hi @michael1011 thanks for your feedback. I addressed your points and added one question about `sign_refund`.

File: src/swaps/bitcoin.rs (hunk 3 above)
Comment (michael1011) [reply src/swaps/bitcoin.rs]: It depends on what part of the transaction you are changing. Here we are just talking about the witness (= unlocking script here: https://wiki.bitcoinsv.io/index.php/SIGHASH_flags#Items_that_are_never_signed), which is not signed. You can ask for partial signatures of either the unsigned or partially signed transaction; both will work just fine

**Comment by i5hi** [comment]
@ok300 Have you tried running the integration submarine tests and send 2 payments to the address to see if this all works well?

**Comment by i5hi** [comment]
Also ensure that standard refunds are working as expected

**Comment by ok300** [comment]
@i5hi I ran `cargo test` and some tests passed, others were skipped with "ignored, requires invoice and refund address".

Are these the tests you mean and if so, what's the proper way to run them? Just change the invoice and refund address fields, then remove the `#ignore` annotation?

**Comment by i5hi** [comment]
you dont need to remove ignore, for each submarine test you have to use a fresh invoice or you will get a 400 from boltz:

```
cargo test --test bitcoin bitcoin_v2_submarine -- --nocapture --ignored 
```

**Comment by ok300** [comment]
I tried `bitcoin_v2_submarine` with a fresh testnet invoice from Phoenix, but running the command you posted failed with "invoice not for current active network: testnet3".

Do I have to set anything else?

**Comment by i5hi** [comment]
Just tried a pheonix testnet invoice; worked fine.

Try this

```
lntb5125180n1pnwmt4zpp5mvhqku7k4m4smvhdft67qq0kfpsg3jktsp99uaws60rq6m9wumfqcqpjsp5qlw0j3wlpnzx4cmrk3n8f09fqra2fmwjm4jk0j2l8uqeclag8c9q9q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqdqqmqz9gxqyjw5qrzjqwfn3p9278ttzzpe0e00uhyxhned3j5d9acqak5emwfpflp8z2cnflctr6qq3f9n3gqqqqlgqqqqqeqqjqeh4hcz22hnr9xc0tat79q0aujefdreewpseqvh6z4afk2qyd067r5z8zq863m5lgkj9nkrdmmtalf6j7zejla0xur3e22mq4dtev9tqpkhanae
```

**Comment by ok300** [comment]
I ran two kinds of tests:

First, I used this branch to trigger (and complete) refunds for Incoming Chain Swaps (BTC -> LBTC) in https://github.com/breez/breez-sdk-liquid/pull/471 . Both cooperative and non-cooperative claims (to LBTC, as final step of the chain swap) and refunds (to BTC) were successful.

Second, I used @i5hi 's suggestion above to run the `bitcoin_v2_submarine` test. It essentially worked (the new invoice was paid), even though the test printed an error `HTTP("mempool min fee not met, 224 < 4543")` at L170, which is when calling `boltz_api_v2.post_submarine_claim_tx_details`. The testnet mempool fee estimation is generally erratic. I also didn't find a way to set custom (higher) feerates for the claim tx in the `bitcoin_v2_submarine` test. IMO this is a separate issue related to feerates on testnet.

The PR is ready for final review 🙏

**Comment by ok300** [comment]
@michael1011 @i5hi anything else I can do to help with the review?

We are using this branch in the Liquid Breez SDK and it would make life easier to get it merged 🙏

**Review by michael1011 — APPROVED** [review]
utACK

**Comment by i5hi** [comment]
Thanks @ok300

**Pull request merged**